futures/src/error.rs

```diff
@@ -1,16 +1,144 @@
-use std::fmt;
 use failure::{Backtrace, Context, Fail};
+use lapin_async;
+use std::fmt;
+use std::io;
+use tokio_timer;
+
+use transport::CodecError;
 
 /// The type of error that can be returned in this crate.
+///
+/// Instead of implementing the `Error` trait provided by the standard library,
+/// it implemented the `Fail` trait provided by the `failure` crate. Doing so
+/// means that this type guaranteed to be both sendable and usable across
+/// threads, and that you'll be able to use the downcasting feature of the
+/// `failure::Error` type.
 #[derive(Debug)]
 pub struct Error {
-    inner: Context<String>,
+    inner: Context<ErrorKind>,
+}
+
+/// The different kinds of errors that can be reported.
+///
+/// This enumeration is deliberately not exported to the end-users of this
+/// crate because it is not yet possible to prevent developers of matching
+/// exhaustively against its variants.
+#[derive(Debug, Fail)]
+pub(crate) enum ErrorKind {
+    #[fail(display = "The maximum number of channels for this connection has been reached")]
+    ChannelLimitReached,
+    #[fail(display = "Failed to open channel")]
+    ChannelOpenFailed,
+    #[fail(display = "Couldn't decode incoming frame: {}", _0)]
+    Decode(CodecError),
+    #[fail(display = "The connection was closed by the remote peer")]
+    ConnectionClosed,
+    #[fail(display = "Failed to connect: {}", _0)]
+    ConnectionFailed(#[fail(cause)] io::Error),
+    #[fail(display = "Basic get returned empty")]
+    EmptyBasicGet,
+    #[fail(display = "Couldn't encode outcoming frame: {}", _0)]
+    Encode(CodecError),
+    #[fail(display = "The timer of the heartbeat encountered an error: {}", _0)]
+    HeartbeatTimer(#[fail(cause)] tokio_timer::Error),
+    #[fail(display = "Failed to handle incoming frame: {:?}", _0)]
+    // FIXME: mark lapin_async's Error as cause once it implements Fail
+    InvalidFrame(lapin_async::error::Error),
+    #[fail(display = "Couldn't parse URI: {}", _0)]
+    InvalidUri(String),
+    #[fail(display = "Transport mutex is poisoned")]
+    PoisonedMutex,
+    #[fail(display = "{}: {:?}", _0, _1)]
+    // FIXME: mark lapin_async's Error as cause once it implements Fail
+    ProtocolError(String, lapin_async::error::Error)
 }
 
 impl Error {
-    pub(crate) fn new(message: impl Into<String>) -> Self {
-        Error {
-            inner: Context::new(message.into())
+    /// Returns true if the error is caused by the limit of channels being reached.
+    pub fn is_channel_limit_reached(&self) -> bool {
```

Comment:
Rather than having several boolean methods like this, would it make sense to just make ErrorKind public and just return the ErrorKind?

Reply:
As much as I'd like to, doing so would prevent us from adding new error variants once this library is considered stable: if any people were to match exhaustively match against the variants, as soon as we add one their code breaks. There's an accepted RFC for this feature, unfortunately it hasn't been [implemented yet](https://github.com/rust-lang/rust/issues/44109). I'd be happy to expose the `ErrorKind` enum once it's been stabilized though.

Reply:
Maybe we could add some « Other(String) » ErrorKind, and when/if we add some new one, juste use Other for it with a FIXME comment explaining the problem. Then later on make a breaking pass adding the new real kinds in the enum when we feel it’s ok to break this part for the specific case of people doing exhaustive match on this

Reply:
Another workaround that has been used in the standard library before they used the `#[non_exhaustive]` attribute (it would seem that it's already implemented, but not stabilized), is to have a [hidden `__Nonexhaustive` variant](https://doc.rust-lang.org/1.0.0/src/std/io/error.rs.html#118-123), to prevent developers from exhaustively matching the enumeration, and still allow adding new variants to the enumeration.

Reply:
I’m fine with the workaround and a comment explaining why, then!